### PR TITLE
NitroxTransform reworked

### PR DESCRIPTION
**Pull request by killzoms**
swapped array usage with explicit fields

### CODE REVIEW

**Review by dartasen — APPROVED** [review]

**Comment by Measurity** [comment]
This is a bit tricky to review and verify. It seems fine having some unit tests wouldn't be out of place before getting merged.

**Comment by killzoms** [comment]
Notice:

Something cropped up with one of the tests resulting in a larger drift of values than expected, locating source and solving

**Comment by killzoms** [comment]
Issue with Quaternion ToEuler/FromEuler solved, just left with composing and decomposing the Rotation Matrix

**Comment by killzoms** [comment]
everything appears to work now

**Comment by killzoms** [comment]
update: Composition works Decomposition broken again

**Comment by killzoms** [comment]
Alright Composition and Decomposition produce predicted results, however Singularities are not taken into account

File: NitroxModel/DataStructures/GameLogic/NitroxQuaternion.cs [hunk 1]
```diff
@@ -146,15 +146,16 @@ private static NitroxVector3 NormalizeAngles(NitroxVector3 angles)
...
 
-            while (angle < 0)
+            while (angle >= 360)
             {
-                angle += 360;
+                angle -= 360;
             }
+
             return angle;
         }
```
Comment (Measurity) [inline NitroxModel/DataStructures/GameLogic/NitroxQuaternion.cs]: Why not use modulus operator here? `return angle % 360;`

Comment (killzoms) [reply NitroxModel/DataStructures/GameLogic/NitroxQuaternion.cs]: does modulus work like that?

**Comment by killzoms** [comment]
alright everything APPEARS to be working test wise

File: NitroxModel/DataStructures/GameLogic/NitroxMatrix4x4.cs [hunk 2]
```diff
@@ -313,53 +313,143 @@ public static NitroxVector3 GetScale(ref NitroxMatrix4x4 matrix)
             return scale;
         }
 
-        // From https://github.com/Unity-Technologies/Unity.Mathematics/blob/4915b7afebc50b9c6c9a410b7a86ae5489aa6b9c/src/Unity.Mathematics/matrix.cs#L791
-        private static NitroxMatrix4x4 GetRotationMatrix(NitroxQuaternion rotation)
+        // From https://www.euclideanspace.com/maths/geometry/rotations/conversions/quaternionToMatrix/index.htm
+        public static NitroxMatrix4x4 GetRotationMatrix(NitroxQuaternion rotation)
```
Comment (Jannify) [inline NitroxModel/DataStructures/GameLogic/NitroxMatrix4x4.cs:317]: Would set this `private` like the other GetMatrix methods

File: NitroxModel/DataStructures/GameLogic/NitroxMatrix4x4.cs [hunk 3]
```diff
@@ -313,53 +313,143 @@ public static NitroxVector3 GetScale(ref NitroxMatrix4x4 matrix)
...
+            rotMat.M12 = 2f * (yz - xw);
+
+            return rotMat;
         }
 
-        // From https://answers.unity.com/questions/402280/how-to-decompose-a-trs-matrix.html
-        public static NitroxQuaternion GetRotation(ref NitroxMatrix4x4 matrix)
+        // From https://www.euclideanspace.com/maths/geometry/rotations/conversions/matrixToQuaternion/index.htm
+        // left in case weirdness ensues with the simpler method
+        public static NitroxQuaternion GetRotation2(ref NitroxMatrix4x4 matrix)
```
Comment (Jannify) [inline NitroxModel/DataStructures/GameLogic/NitroxMatrix4x4.cs:360]: Would make the method private so no other class uses this unintentionally.

File: NitroxModel/DataStructures/GameLogic/NitroxQuaternion.cs [hunk 4]
```diff
@@ -22,6 +22,13 @@ public struct NitroxQuaternion
 
         public static NitroxQuaternion Identity { get; } = new NitroxQuaternion(0, 0, 0, 1);
 
+        public bool Normalized {
```
Comment (Jannify) [inline NitroxModel/DataStructures/GameLogic/NitroxQuaternion.cs:25]:
```suggestion
        public bool IsNormalized {
```

File: NitroxModel/DataStructures/GameLogic/NitroxVector3.cs [hunk 5]
```diff
@@ -80,20 +88,39 @@ public override int GetHashCode()
...
+        public static NitroxVector3 operator /(NitroxVector3 lhs, double rhs)
+        {
+            return new((float)(lhs.X / rhs), (float)(lhs.Y / rhs), (float)(lhs.Z / rhs));
+        }
+
+        public static NitroxVector3 operator *(NitroxVector3 lhs, double rhs)
+        {
+            return new((float)(lhs.X * rhs), (float)(lhs.Y * rhs), (float)(lhs.Z * rhs));
+        }
+
```
Comment (Jannify) [inline NitroxModel/DataStructures/GameLogic/NitroxVector3.cs:100]: Afaik they are unused and can be removed

File: NitroxModel/DataStructures/GameLogic/NitroxVector3.cs [hunk 6]
```diff
@@ -80,20 +88,39 @@ public override int GetHashCode()
...
         {
             return !(lhs == rhs);
         }
 
+        public static implicit operator NitroxVector3(NitroxVector4 vector4)
+        {
+            return new NitroxVector3(vector4.X, vector4.Y, vector4.Z);
+        }
+
         public static NitroxVector3 Normalize(NitroxVector3 value)
```
Comment (Jannify) [inline NitroxModel/DataStructures/GameLogic/NitroxVector3.cs:116]:
```suggestion

        // Copied from Unity
        public static NitroxVector3 Normalize(NitroxVector3 value)
```

**Review by Jannify — CHANGES_REQUESTED** [review]

**Comment by killzoms** [comment]
Implemented elsewhere

**Pull request closed without merging**